### PR TITLE
rpcdaemon: StateFactory rework

### PR DESCRIPTION
**Pull request by lupin012**
This PR contains:
- removed StateFactory::create_state_txn()
- modified StateFactory::create_state() uses txn_id as input no more block_num
- StateFactory::user_txn_id_at() new method that return txn_id from block_num and txn_index
- modify APIs using new StateFactory interface

### CODE REVIEW

File: silkworm/rpc/core/receipts.cpp [hunk 1]
```diff
@@ -152,10 +152,14 @@ Task<std::optional<Receipts>> generate_receipts(db::kv::api::Transaction& tx, co
 
     const auto chain_config = co_await chain_storage.read_chain_config();
     auto current_executor = co_await boost::asio::this_coro::executor;
+
+    execution::StateFactory state_factory{tx};
+    auto txn_id = co_await state_factory.get_txn_id(block_num);
```
Comment (canepat) [inline silkworm/rpc/core/receipts.cpp:157]: `txn_id` should be `const`

File: silkworm/rpc/core/evm_trace.cpp [hunk 2]
```diff
@@ -1795,14 +1812,12 @@ Task<TraceCallResult> TraceCallExecutor::execute(
...
-    std::shared_ptr<State> curr_state{};
     // We must do the execution at the state after the txn identified by the given index within the given block
-    // at the state after the block identified by the given block_num, i.e. at the start of the next block (block_num + 1)
-    const auto first_txn_num_in_next_block = co_await tx_.first_txn_num_in_block(block_num + 1);
-    const auto txn_id = first_txn_num_in_next_block + 1 + gsl::narrow<TxnId>(transaction.transaction_index);  // + 1 for system txn in the beginning of block
-    state = execution::StateFactory{tx_}.create_state_txn(current_executor, chain_storage_, txn_id);
-    curr_state = execution::StateFactory{tx_}.create_state_txn(current_executor, chain_storage_, txn_id);
+    // at the state after the block identified by the given block_num,
+    execution::StateFactory state_factory{tx_};
+    auto txn_id = co_await state_factory.get_txn_id(block_num, gsl::narrow<uint32_t>(transaction.transaction_index));
```
Comment (canepat) [inline silkworm/rpc/core/evm_trace.cpp:1818]: `txn_id` should be `const`

File: silkworm/rpc/core/evm_trace.cpp [hunk 3]
```diff
@@ -1697,11 +1711,14 @@ Task<bool> TraceCallExecutor::trace_touch_block(const silkworm::BlockWithHash& b
 
     const auto chain_config = co_await chain_storage_.read_chain_config();
     auto current_executor = co_await boost::asio::this_coro::executor;
+    execution::StateFactory state_factory{tx_};
+    auto txn_id = co_await state_factory.get_txn_id(block_num);
```
Comment (canepat) [inline silkworm/rpc/core/evm_trace.cpp:1715]: `txn_id` should be `const`

File: silkworm/rpc/core/evm_trace.cpp [hunk 4]
```diff
@@ -1667,14 +1681,14 @@ Task<TraceOperationsResult> TraceCallExecutor::trace_operations(const Transactio
 
     // We must do the execution at the state after the txn identified by transaction_with_block param in the same block
     // at the state of the block identified by the given block_num, i.e. at the start of the block (block_num)
-    const auto first_txn_num_in_block = co_await tx_.first_txn_num_in_block(block_num);
-    const auto trace_op_result = co_await async_task(workers_.executor(), [&]() -> TraceOperationsResult {
-        const auto txn_id = first_txn_num_in_block + 1 + static_cast<uint64_t>(transaction_with_block.transaction.transaction_index);  // + 1 for system txn in the beginning of block
+    execution::StateFactory state_factory{tx_};
+    auto txn_id = co_await state_factory.get_txn_id(block_num, gsl::narrow<uint32_t>(transaction_with_block.transaction.transaction_index));
```
Comment (canepat) [inline silkworm/rpc/core/evm_trace.cpp:1685]: `txn_id` should be `const`

File: silkworm/rpc/core/evm_trace.cpp [hunk 5]
```diff
@@ -1634,14 +1648,14 @@ Task<std::string> TraceCallExecutor::trace_transaction_error(const TransactionWi
 
     // We must do the execution at the state after the txn identified by transaction_with_block param in the same block
     // at the state of the block identified by the given block_num, i.e. at the start of the block (block_num)
-    const auto first_txn_num_in_block = co_await tx_.first_txn_num_in_block(block_num);
-    const auto trace_error = co_await async_task(workers_.executor(), [&]() -> std::string {
-        const auto txn_id = first_txn_num_in_block + 1 + static_cast<uint64_t>(transaction_with_block.transaction.transaction_index);  // + 1 for system txn in the beginning of block
+    execution::StateFactory state_factory{tx_};
+    auto txn_id = co_await state_factory.get_txn_id(block_num, gsl::narrow<uint32_t>(transaction_with_block.transaction.transaction_index));
```
Comment (canepat) [inline silkworm/rpc/core/evm_trace.cpp:1652]: `txn_id` should be `const`

File: silkworm/rpc/core/evm_trace.cpp [hunk 6]
```diff
@@ -1605,13 +1619,13 @@ Task<TraceEntriesResult> TraceCallExecutor::trace_transaction_entries(const Tran
 
     // We must do the execution at the state after the txn identified by transaction_with_block param in the same block
     // at the state of the block identified by the given block_num, i.e. at the start of the block (block_num)
-    const auto first_txn_num_in_block = co_await tx_.first_txn_num_in_block(block_num);
-    const auto trace_result = co_await async_task(workers_.executor(), [&]() -> TraceEntriesResult {
-        const auto txn_id = first_txn_num_in_block + 1 + static_cast<uint64_t>(transaction_with_block.transaction.transaction_index);  // + 1 for system txn in the beginning of block
+    execution::StateFactory state_factory{tx_};
+    auto txn_id = co_await state_factory.get_txn_id(block_num, gsl::narrow<uint32_t>(transaction_with_block.transaction.transaction_index));
```
Comment (canepat) [inline silkworm/rpc/core/evm_trace.cpp:1623]: `txn_id` should be `const`

File: silkworm/rpc/core/evm_trace.cpp [hunk 7]
```diff
@@ -1540,11 +1550,15 @@ Task<TraceDeployResult> TraceCallExecutor::trace_deploy_transaction(const silkwo
 
     const auto chain_config = co_await chain_storage_.read_chain_config();
     auto current_executor = co_await boost::asio::this_coro::executor;
+
+    execution::StateFactory state_factory{tx_};
+    auto txn_id = co_await state_factory.get_txn_id(block_num);
```
Comment (canepat) [inline silkworm/rpc/core/evm_trace.cpp:1555]: `txn_id` should be `const`

File: silkworm/rpc/core/evm_trace.cpp [hunk 8]
```diff
@@ -1479,12 +1484,17 @@ Task<TraceManyCallResult> TraceCallExecutor::trace_calls(const silkworm::Block&
 
     const auto chain_config = co_await chain_storage_.read_chain_config();
     auto current_executor = co_await boost::asio::this_coro::executor;
+
+    execution::StateFactory state_factory{tx_};
+    // trace_calls semantics: we must execute the call from the state at the end of the given block, so we pass block.header.number + 1
+    auto txn_id = co_await state_factory.get_txn_id(block_num + 1);
```
Comment (canepat) [inline silkworm/rpc/core/evm_trace.cpp:1490]: `txn_id` should be `const`

File: silkworm/rpc/core/evm_trace.cpp [hunk 9]
```diff
@@ -1413,14 +1413,19 @@ Task<std::vector<TraceCallResult>> TraceCallExecutor::trace_block_transactions(c
 
     const auto chain_config = co_await chain_storage_.read_chain_config();
     auto current_executor = co_await boost::asio::this_coro::executor;
+
+    execution::StateFactory state_factory{tx_};
+    // trace_block semantics: we must execute the call from the state at the current block
+    auto txn_id = co_await state_factory.get_txn_id(block_num);
```
Comment (canepat) [inline silkworm/rpc/core/evm_trace.cpp:1419]: `txn_id` should be `const`

File: silkworm/rpc/core/evm_debug.cpp [hunk 10]
```diff
@@ -551,12 +555,11 @@ Task<void> DebugExecutor::execute(
 
     // We must do the execution at the state after the txn identified by transaction_with_block param in the same block
     // at the state of the block identified by the given block_num, i.e. at the start of the block (block_num)
-    const auto first_txn_num_in_block = co_await tx_.first_txn_num_in_block(block.header.number + 1);
+    execution::StateFactory state_factory{tx_};
+    auto txn_id = co_await state_factory.get_txn_id(block.header.number, static_cast<uint32_t>(transaction_index));
```
Comment (canepat) [inline silkworm/rpc/core/evm_debug.cpp:559]: `txn_id` should be `const`

File: silkworm/rpc/core/evm_debug.cpp [hunk 11]
```diff
@@ -434,8 +434,12 @@ Task<void> DebugExecutor::execute(json::Stream& stream, const ChainStorage& stor
 
     const auto chain_config = co_await storage.read_chain_config();
     auto current_executor = co_await boost::asio::this_coro::executor;
+
+    execution::StateFactory state_factory{tx_};
+    auto txn_id = co_await state_factory.get_txn_id(block_num);
```
Comment (canepat) [inline silkworm/rpc/core/evm_debug.cpp:439]: `txn_id` should be `const`

File: silkworm/rpc/commands/eth_api.cpp [hunk 12]
```diff
@@ -1436,9 +1446,12 @@ Task<void> EthereumRpcApi::handle_eth_call_bundle(const nlohmann::json& request,
                 break;
             }
 
+            execution::StateFactory state_factory{*tx};
+            auto txn_id = co_await state_factory.get_txn_id(block_with_hash->block.header.number + 1);
```
Comment (canepat) [inline silkworm/rpc/commands/eth_api.cpp:1450]: `txn_id` should be `const`

File: silkworm/rpc/commands/eth_api_test.cpp [hunk 13]
```diff
@@ -105,6 +105,7 @@ TEST_CASE_METHOD(test_util::RpcApiE2ETest, "unit: eth_feeHistory succeeds if req
     })"_json);
 }
 
+#ifdef notdef  // commented Temporary waiting implementaion local-transaction/
```
Comment (canepat) [inline silkworm/rpc/commands/eth_api_test.cpp:108]:
```suggestion
#ifdef notdef  // temporarily commented out waiting for LocalTransaction implementation
```

File: silkworm/rpc/commands/rpc_api_test.cpp [hunk 14]
```diff
@@ -24,7 +24,7 @@
 
 namespace silkworm::rpc::commands {
 
-#ifdef notdef  // commented Temporary wating implementaion local-transaction
+#ifdef notdef  // commented Temporary waiting implementaion local-transaction
```
Comment (canepat) [inline silkworm/rpc/commands/rpc_api_test.cpp:27]:
```suggestion
#ifdef notdef  // temporarily commented out waiting for LocalTransaction implementation
```

File: silkworm/rpc/commands/rpc_api_test.cpp [hunk 15]
```diff
@@ -78,7 +78,7 @@ static const std::vector<std::string> kSubtestsToIgnore = {
     "call-simple-contract.io",      // eth_call: without gas paramters doesn't support base_fee_gas of block as default gas
 };
 
-#ifdef notdef  // commented Temporary wating implementaion local-transaction
+#ifdef notdef  // commented Temporary waiting implementaion local-transaction
```
Comment (canepat) [inline silkworm/rpc/commands/rpc_api_test.cpp:81]:
```suggestion
#ifdef notdef  // temporarily commented out waiting for LocalTransaction implementation
```

**Review by canepat — CHANGES_REQUESTED** [review]

File: silkworm/execution/state_factory.hpp [hunk 16]
```diff
@@ -31,14 +31,14 @@ struct StateFactory {
...
-        boost::asio::any_io_executor& executor,
-        const db::chain::ChainStorage& storage,
-        BlockNum block_num);
-
-    std::shared_ptr<State> create_state_txn(
         boost::asio::any_io_executor& executor,
         const db::chain::ChainStorage& storage,
         TxnId txn_id);
+
+    Task<TxnId> get_txn_id(BlockNum block_num, uint32_t tx_id = 0) {
```
[the remark this answers is not in the document]
Comment (lupin012) [reply silkworm/execution/state_factory.hpp:38]: ok

File: silkworm/rpc/commands/eth_api.cpp [hunk 17]
```diff
@@ -923,8 +923,11 @@ Task<void> EthereumRpcApi::handle_eth_estimate_gas(const nlohmann::json& request
             co_return co_await state_reader.read_account(address);
         };
 
+        execution::StateFactory state_factory{*tx};
+        auto txn_id = co_await state_factory.get_txn_id(latest_block.header.number);
```
[the remark this answers is not in the document]
Comment (lupin012) [reply silkworm/rpc/commands/eth_api.cpp:927]: ok

File: silkworm/rpc/core/receipts.cpp (hunk 1 above)
Comment (lupin012) [reply silkworm/rpc/core/receipts.cpp:157]: ok

File: silkworm/rpc/commands/eth_api.cpp [hunk 18]
```diff
@@ -1154,9 +1157,12 @@ Task<void> EthereumRpcApi::handle_eth_call(const nlohmann::json& request, std::s
         }
         silkworm::Transaction txn{call.to_transaction()};
 
+        execution::StateFactory state_factory{*tx};
+        auto txn_id = co_await state_factory.get_txn_id(block_num + 1);
```
[the remark this answers is not in the document]
Comment (lupin012) [reply silkworm/rpc/commands/eth_api.cpp:1161]: ok

File: silkworm/rpc/commands/eth_api.cpp [hunk 19]
```diff
@@ -1340,10 +1346,14 @@ Task<void> EthereumRpcApi::handle_eth_create_access_list(const nlohmann::json& r
         Tracers tracers{tracer};
         auto txn = call.to_transaction(std::nullopt, nonce);
         AccessList saved_access_list = call.access_list;
+
+        execution::StateFactory state_factory{*tx};
+        auto txn_id = co_await state_factory.get_txn_id(block_with_hash->block.header.number + 1);
```
[the remark this answers is not in the document]
Comment (lupin012) [reply silkworm/rpc/commands/eth_api.cpp:1351]: ok

File: silkworm/rpc/core/evm_debug.cpp (hunk 11 above)
Comment (lupin012) [reply silkworm/rpc/core/evm_debug.cpp:441]: ok

File: silkworm/rpc/commands/eth_api_test.cpp (hunk 13 above)
Comment (lupin012) [reply silkworm/rpc/commands/eth_api_test.cpp:108]: ok

File: silkworm/rpc/commands/eth_api.cpp (hunk 12 above)
Comment (lupin012) [reply silkworm/rpc/commands/eth_api.cpp:1450]: ok

File: silkworm/rpc/commands/rpc_api_test.cpp (hunk 15 above)
Comment (lupin012) [reply silkworm/rpc/commands/rpc_api_test.cpp:81]: ok

File: silkworm/rpc/commands/rpc_api_test.cpp (hunk 14 above)
Comment (lupin012) [reply silkworm/rpc/commands/rpc_api_test.cpp:27]: ok

**Review by canepat — APPROVED** [review]

**Pull request merged**